### PR TITLE
Linters for trailing blank lines and extra blank lines

### PR DESCRIPTION
**Pull request by printercu**
Hi!

Here are linters similar to rubocop's ones.

### CODE REVIEW

File: lib/slim_lint/linter/empty_lines.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module SlimLint
+  # Checks for trailing whitespace.
```
Comment (sds) [inline lib/slim_lint/linter/empty_lines.rb]: This comment needs to be updated.

File: lib/slim_lint/linter/trailing_blank_lines.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module SlimLint
+  # Checks for trailing whitespace.
```
Comment (sds) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: This comment needs to be updated.

File: lib/slim_lint/linter/empty_lines.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  class Linter::EmptyLines < Linter
+    include LinterRegistry
+
+    on_start do |_sexp|
+      dummy_node = Struct.new(:line)
+
+      was_empty = true
+      document.source.lines.each_with_index do |line, i|
```
Comment (sds) [inline lib/slim_lint/linter/empty_lines.rb]: Instead of calling `document.source.lines`, you can call `document.source_lines` directly.

File: lib/slim_lint/linter/trailing_blank_lines.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  class Linter::TrailingBlankLines < Linter
+    include LinterRegistry
+
+    on_start do |_sexp|
+      dummy_node = Struct.new(:line)
+
+      if !document.source.end_with?("\n")
+        report_lint(dummy_node.new(document.source_lines.size),
+                    'No blank line in the end of file')
+      elsif document.source.lines.last.blank?
```
Comment (sds) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: `document.source.lines` should use `document.source_lines` so you aren't re-splitting the string for newlines here and on line 13 below.

File: lib/slim_lint/linter/trailing_blank_lines.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        report_lint(dummy_node.new(document.source.lines.size),
+                    'Multiple empty lines in the end of file')
```
Comment (sds) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: Shouldn't this case be handled by the `EmptyLines` linter? Why are we duplicating the check?

In RuboCop, they have a `Style/FinalNewline` linter, which is what I think you want to emulate here, since you simply want to ensure whether or not the file has a trailing newline.

File: spec/support/normalize_indent.rb [hunk 6]
```diff
@@ -2,8 +2,8 @@ module IndentNormalizer
   # Strips off excess leading indentation from each line so we can use Heredocs
   # for writing code without having the leading indentation count.
   def normalize_indent(code)
-    leading_indent = code[/^(\s*)/, 1]
-    code.lstrip.gsub(/\n#{leading_indent}/, "\n")
+    leading_indent = code[/^(\s*?)(\n|\S)/, 1]
+    code.gsub(/^#{leading_indent}/, '')
```
Comment (sds) [inline spec/support/normalize_indent.rb]: Why was this changed? It's not clear to me if you are changing behavior or simply accomplishing the same task a different way.

File: lib/slim_lint/linter/empty_lines.rb (hunk 3 above)
Comment (printercu) [inline lib/slim_lint/linter/empty_lines.rb]: `source_lines` are stripped already. There is test for it, and it will fail.

File: spec/support/normalize_indent.rb (hunk 6 above)
Comment (printercu) [inline spec/support/normalize_indent.rb]: It was not working as really intended, and got an issue with first blank line. Seems like I was not able to test warning when first line is blank, because it was stripped.

File: lib/slim_lint/linter/trailing_blank_lines.rb (hunk 5 above)
Comment (printercu) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: This one checks if line is blank (can be not empty but contain only spaces). I don't see this cop you mentioned in current rubocop repo. Seems like it was merged into https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/layout/trailing_blank_lines.rb#L6

File: lib/slim_lint/linter/trailing_blank_lines.rb (hunk 4 above)
Comment (printercu) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: Same as above. `source_lines` strips spaces.

File: lib/slim_lint/linter/empty_lines.rb (hunk 3 above)
Comment (printercu) [inline lib/slim_lint/linter/empty_lines.rb]:
```
$ irb
2.3.3 :001 > "\n\n".split("\n")
```

File: spec/support/normalize_indent.rb (hunk 6 above)
Comment (printercu) [inline spec/support/normalize_indent.rb]:
```
> /^(\s*)/.match " \n "
 => #<MatchData " \n " 1:" \n ">
```

**Comment by printercu** [comment]
@sds thank you. I've missed that copied description for new linters. Updated commits.

File: lib/slim_lint/linter/trailing_blank_lines.rb (hunk 5 above)
Comment (printercu) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: Sorry, I haven't understand question for the first time. This linters are different, and people may want to use them separately. Ex., they may want to disable EmptyLines to divide blocks with 2 or more empty lines, but still want to check to for single trailing blank line.

Comment (printercu) [inline lib/slim_lint/linter/trailing_blank_lines.rb]: And i've just made in the similar way to rubocop.

**Comment by sds** [comment]
Thanks for clarifying, @printercu. I didn't realize RuboCop renamed its `FinalNewline` cop. What you have makes sense here. Thanks!
